src/node_crypto.cc

```diff
@@ -211,6 +211,95 @@ static int NoPasswordCallback(char* buf, int size, int rwflag, void* u) {
 }
 
 
+namespace error {
+Local<Object> DecorateO(Environment* env, Local<Object> obj, unsigned long err) {
+  if (err == 0) return obj;  // No decoration possible.
+
+  const char* ls = ERR_lib_error_string(err);
+  const char* fs = ERR_func_error_string(err);
+  const char* rs = ERR_reason_error_string(err);
+
+  obj->Set(env->isolate()->GetCurrentContext(),
+      OneByteString(env->isolate(), "library"),
+      OneByteString(env->isolate(), ls ? ls : "")).FromJust();
+  obj->Set(env->isolate()->GetCurrentContext(),
+      OneByteString(env->isolate(), "function"),
+      OneByteString(env->isolate(), fs ? fs : "")).FromJust();
+  obj->Set(env->isolate()->GetCurrentContext(),
+      env->reason_string(),
+      OneByteString(env->isolate(), rs ? rs : "")).FromJust();
+
+  // SSL has no API to recover the error name from the number, so we
+  // transform "this error" to "ERR_SSL_THIS_ERROR", which ends up being
+  // close to the original error name.
+  std::string code(rs);
+
+  for (auto& c : code) {
+    if (c == ' ')
+      c = '_';
+    else
+      c = ::toupper(c);
+  }
+
+  const char* prefix = "ERR_NONE_";
+  switch(ERR_GET_LIB(err)) {
+    case ERR_LIB_SYS:    prefix = "ERR_SYS_"; break;
+    case ERR_LIB_BN:     prefix = "ERR_BN_"; break;
+    case ERR_LIB_RSA:    prefix = "ERR_RSA_"; break;
+    case ERR_LIB_DH:     prefix = "ERR_DH_"; break;
+    case ERR_LIB_EVP:    prefix = "ERR_EVP_"; break;
+    case ERR_LIB_BUF:    prefix = "ERR_BUF_"; break;
+    case ERR_LIB_OBJ:    prefix = "ERR_OBJ_"; break;
+    case ERR_LIB_PEM:    prefix = "ERR_PEM_"; break;
+    case ERR_LIB_DSA:    prefix = "ERR_DSA_"; break;
+    case ERR_LIB_X509:   prefix = "ERR_X509_"; break;
+    case ERR_LIB_ASN1:   prefix = "ERR_ASN1_"; break;
+    case ERR_LIB_CONF:   prefix = "ERR_CONF_"; break;
+    case ERR_LIB_CRYPTO: prefix = "ERR_CRYPTO_"; break;
+    case ERR_LIB_EC:     prefix = "ERR_EC_"; break;
+    case ERR_LIB_SSL:    prefix = "ERR_SSL_"; break;
+    case ERR_LIB_BIO:    prefix = "ERR_BIO_"; break;
+    case ERR_LIB_PKCS7:  prefix = "ERR_PKCS7_"; break;
+    case ERR_LIB_X509V3: prefix = "ERR_X509V3_"; break;
+    case ERR_LIB_PKCS12: prefix = "ERR_PKCS12_"; break;
+    case ERR_LIB_RAND:   prefix = "ERR_RAND_"; break;
+    case ERR_LIB_DSO:    prefix = "ERR_DSO_"; break;
+    case ERR_LIB_ENGINE: prefix = "ERR_ENGINE_"; break;
+    case ERR_LIB_OCSP:   prefix = "ERR_OCSP_"; break;
+    case ERR_LIB_UI:     prefix = "ERR_UI_"; break;
+    case ERR_LIB_COMP:   prefix = "ERR_COMP_"; break;
+    case ERR_LIB_ECDSA:  prefix = "ERR_ECDSA_"; break;
+    case ERR_LIB_ECDH:   prefix = "ERR_ECDH_"; break;
+    case ERR_LIB_OSSL_STORE:  prefix = "ERR_STORE_"; break;
```

Comment:
nit: if we're going to align all `prefix = ...` lines, they should line up with this one

Reply:
fair enough, do we have a preference? They were aligned, but this code predates openssl 1.1.1, which added `OSSL_` into the macro name. Maybe its slightly less pretty but more maintainable to just left-align?

Reply:
I'd go with whatever `make format-cpp` does